keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+|JSON Infof             |1406        |19         |1                    |
+|JSON InfoS             |319         |67         |1                    |
+
+InfoS implementation for text is 9% slower than Infof. This increase should not have a big impact on overall Kubernetes performance as logging takes less than 2% of overall CPU usage.
+JSON implementation using zapr is much more efficient compared to current implementation.
+
+#### Log volume increase analysis
+
+Example log size change based on apiserver HTTP access log. This log generates most volume in kube-apiserver and accepts lots of arguments. This makes this message a good example of the worst case for log volume increase.
+
+```go
+klog.Infof("%s %s: (%v) %v%v%v [%s %s]", rl.req.Method, rl.req.RequestURI, latency, rl.status, rl.statusStack, rl.addedInfo, rl.req.UserAgent(), rl.req.RemoteAddr)
+```
+
+Using old format generates 206 characters (before [changes by @lavalamp](https://github.com/kubernetes/kubernetes/pull/87203))
+```
+I1025 00:15:15.525108       1 httplog.go:79] GET /api/v1/namespaces/kube-system/pods/metrics-server-v0.3.1-57c75779f-9p8wg: (1.512ms) 200 [pod_nanny/v0.0.0 (linux/amd64) kubernetes/$Format 10.56.1.19:51756]
+```
+
+Migrating to InfoS will result in log of length 284 characters [+38%]
+```
+I0129 14:57:28.832349  102767 experiment.go:26] HTTP method=GET uri=/api/v1/namespaces/kube-system/pods/metrics-server-v0.3.1-57c75779f-9p8wg latency=1.512ms status=200 status-stack= added-info= user-agent=”pod_nanny/v0.0.0 (linux/amd64) kubernetes/$Format” remote-addr=10.56.1.19:51756
+```
+
+InfoS with Json enabled will generate log line with 347 characters [+68%]
+```json
+{"level":"info","ts":1580306777.04728,"caller":"zapr@v0.1.1/zapr.go:69","msg":"HTTP","method":"GET","uri":"/api/v1/namespaces/kube-system/pods/metrics-server-v0.3.1-57c75779f-9p8wg","latency":"1.512ms","status":200,"status-stack":"","added-info":"","user-agent":"pod_nanny/v0.0.0 (linux/amd64) kubernetes/$Format","remote-addr":"10.56.1.19:51756"}
+```
+
+Data about byte log length in k8s components. Data taken from scalability test [gce-master-scale-performance](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance)
+
+|component              |average|50%ile|75%ile|90%ile|
+|-----------------------|-------|------|------|------|
+|kube-apiserver         |229    |218   |248   |319   |
+|kube-controller-manager|255    |229   |355   |362   |
+|kubelet                |759    |885   |912   |1143  |
+|kube-scheduler         |217    |225   |226   |227   |
+
+When looking at the whole repository we are not expecting a very big increase of log volume due added metadata. This is due to how big log messages are and low number of arguments passed to formatting messages (average below 2, max 10). Based on this HTTP access logs we are expecting that log volume will not increase by more than 50% for default (text) logging configuration.
```

Comment:
I actually claim that 50% is big increase (given that disc access is actually one of the bottlenecks we're experiencing).

Reply:
This was an upper limit based on worst case example which increased volume by 38%.  
I could work on more accurate estimation that should be lower.

Reply:
I guess what I'm saying is that even if it's 20%, it's non-negligible...

Reply:
It would be hard to propose a log structure improvement without any increase of log volume.  

I think the question should be more "Are proposed mitigation sufficient to address expected increase of log volume?"
Do you think that informing users about expected log volume increase between k8s versions is enough to address it?
We could measure it for new releases and add it in change log.

Reply:
I agree it would be really hard to not increase it. That said I was expecting increase more of like ~10%, than 30-50%...
Would it be possible that we have more format (e.g. somewhat-similar to the old one that doesn't add much, and new-one (being a recommended, potentially also a default one) just to give people visibly more time to prepare (e.g. 3 releases or sth)?

Reply:
I will check if we can target ~10% for log increase by optimizing log lines. Two first log lines are responsible for generating 96% of log lines count. If we can remove some bloat in them (e.g. $Format) that you mention about, we could save enough space to reach target without complicating migration (additional releases, additional format).

Reply:
As serathius mentioned I think that for most of the log entries changing them to structured format should not have a significant impact on the overall logs volume because those logs entries do not occur in logs too often. This is true especially if we will split the migration into several Kubernetes releases.
The problem is with top 5 or top 10 log entries which account for over 90% of all log entries. For those any change like https://github.com/kubernetes/kubernetes/pull/87203 will have a significant impact on overall logs volume.